fix: 'Disable input if prefilled' not working with inline embeds

fiston-user opened the pull request:
## What does this PR do?

The "Disable input if the URL identifier is prefilled" option was not working for inline embed forms. While parameters were correctly loaded into the form, the inputs were not being disabled as expected.

This fix ensures that URL parameters are properly accessible in embed iframe contexts by falling back to window.location.search when useSearchParams doesn't properly reflect the parameters.

- Fixes #23133

## Updates since last revision

**Latest changes (addressing @keith's feedback):**

Refactored to avoid checking the URL inside the hook. Instead, the hook now accepts a `forceWindowLocationParams` option that callers can use to indicate when they need to read directly from window.location:

- `useRouterQuery`: Accepts `forceWindowLocationParams` option instead of checking URL internally
- `useShouldBeDisabledDueToPrefill`: Accepts and passes through the option
- `FormBuilderField`: Accepts and passes through the option
- `BookingFields`: Accepts and passes through the option
- `BookEventForm`: Uses `useIsEmbed()` to determine when to pass `forceWindowLocationParams: true`

**Previous changes (addressing @prashant1177 and @hariombalhara):**

- Simplified the if/else ladder to avoid duplicate `URLSearchParams` parsing
- Used `[].values()` instead of `null` for cleaner fallback
- Added detailed comment explaining why `useSearchParams()` may not work in embed contexts

## Visual Demo (For contributors especially)

#### Image Demo:

**Before Fix:**
- ✅ Input fields are prefilled with values from URL parameters
- ❌ Fields remain editable (not disabled)

**After Fix:**
- ✅ Input fields are prefilled with values from URL parameters  
- ✅ Fields are disabled/read-only when "Disable input if prefilled" is enabled

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A - I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs).
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Are there environment variables that should be set? **No special env vars needed**
- What are the minimal test data to have? **An event type with "Disable input if the URL identifier is prefilled" option enabled**
- What is expected (happy path) to have (input and output)?
  - **Input**: Inline embed with prefilled parameters like `calLink: "username/event?name=John%20Doe&email=john@example.com"`
  - **Output**: The name and email fields should be both prefilled AND disabled (read-only)
- Any other important info that could help to test that PR: **A test HTML file is included in the branch for easy testing**

## Human Review Checklist

- [ ] Verify `forceWindowLocationParams` is correctly passed through the entire component chain (BookEventForm → BookingFields → FormBuilderField → useShouldBeDisabledDueToPrefill → useRouterQuery)
- [ ] Confirm `useIsEmbed()` returns true in inline embed contexts
- [ ] Test the "Disable input if prefilled" feature in an actual inline embed scenario
- [ ] Verify the fix doesn't affect non-embed booking forms

## Checklist

- I have read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code follows the style guidelines of this project
- I have checked if my changes generate no new warnings

---

**Link to Devin run**: https://app.devin.ai/sessions/031202650ac247e19c7ee09aef2372b9
**Requested by**: unknown ()

fiston-user committed:
fix: Enable 'Disable input if prefilled' for inline embeds

The "Disable input if the URL identifier is prefilled" option was not working
for inline embed forms. While parameters were correctly loaded into the form,
the inputs were not being disabled as expected.

This fix ensures that URL parameters are properly accessible in embed iframe
contexts by falling back to window.location.search when useSearchParams
doesn't properly reflect the paramters.

Fixes #23133

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/calcom/cal.com?pullRequest=23135) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/calcom/cal.com?pullRequest=23135) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/calcom/cal.com?pullRequest=23135) it.</sub>

prashant1177 reviewed (changes requested):
Change 1. I think it will be better if we check `[]` instead of null
let routerQuery = fromEntriesWithDuplicateKeys(searchParams?.entries() ?? []);

Change 2. You can get rid of if/else ladder by directly checking  window.location.pathname.includes("/embed") 

In this way you can just do it like avoid declaring urlSearchParams twice.

Final if block could be like this
```

  if (typeof window !== "undefined" && window.location.pathname.includes("/embed")) {
    const urlSearchParams = new URLSearchParams(window.location.search);
    const windowQuery = fromEntriesWithDuplicateKeys(urlSearchParams.entries());

    routerQuery = Object.keys(routerQuery).length === 0
      ? windowQuery
      : { ...routerQuery, ...windowQuery };
  }

```

Devanshusharma2005 reviewed (commented):
lets add a loom video to depict the before and after changes.

kart1ka committed:
Merge branch 'main' into fix/inline-embed-disable-prefilled-input

hariombalhara reviewed (changes requested):
Left comments

devin-ai-integration[bot] committed:
refactor: simplify embed URL params fallback logic

Address review feedback:
- Use [].values() instead of null for cleaner fallback
- Simplify if/else ladder to avoid duplicate URLSearchParams parsing
- Add detailed comment explaining why this workaround is needed

Co-authored-by: fiston-user <fiston-user@users.noreply.github.com>
Co-Authored-By: unknown <>

devin-ai-integration[bot] committed:
refactor: use forceWindowLocationParams parameter instead of URL checking

Addresses Keith's feedback to avoid checking URL for embed inside the hook.
Instead, the hook now accepts a forceWindowLocationParams option that callers
can use to indicate when they need to read directly from window.location.

Changes:
- useRouterQuery: Accept forceWindowLocationParams option instead of checking URL
- useShouldBeDisabledDueToPrefill: Accept and pass through forceWindowLocationParams
- FormBuilderField: Accept and pass through forceWindowLocationParams
- BookingFields: Accept and pass through forceWindowLocationParams
- BookEventForm: Pass forceWindowLocationParams based on isEmbed context

Co-authored-by: fiston-user <fiston-user@users.noreply.github.com>
Co-Authored-By: unknown <>
